// clashofclans.js/dist/struct/Clan.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Clan = void 0;
const Decorators_1 = require("../util/Decorators");
const Badge_1 = require("./Badge");
const ClanCapital_1 = require("./ClanCapital");
const ClanMember_1 = require("./ClanMember");
const Location_1 = require("./Location");
/** Represents a Clan. */
class Clan {
    constructor(client, data) {
        this.client = client;
        this.name = data.name;
        this.tag = data.tag;
        this.type = data.type;
        this.description = data.description;
        this.location = data.location ? new Location_1.Location(data.location) : null;
        this.chatLanguage = data.chatLanguage ?? null;
        this.badge = new Badge_1.Badge(data.badgeUrls);
        this.level = data.clanLevel;
        this.points = data.clanPoints;
        this.builderBasePoints = data.clanBuilderBasePoints;
        this.requiredTrophies = data.requiredTrophies;
        this.requiredTownHallLevel = data.requiredTownhallLevel ?? null;
        this.requiredBuilderBaseTrophies = data.requiredBuilderBaseTrophies ?? null;
        this.warFrequency = data.warFrequency;
        this.warWinStreak = data.warWinStreak;
        this.warWins = data.warWins;
        this.warTies = data.warTies ?? null;
        this.warLosses = data.warLosses ?? null;
        this.isWarLogPublic = data.isWarLogPublic;
        this.warLeague = data.warLeague ?? null;
        this.memberCount = data.members;
        this.labels = data.labels;
        // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
        this.clanCapital = Object.keys(data.clanCapital ?? {}).length > 0 ? new ClanCapital_1.ClanCapital(data.clanCapital) : null;
        this.isFamilyFriendly = data.isFamilyFriendly;
        this.capitalPoints = data.clanCapitalPoints;
        this.capitalLeague = data.capitalLeague;
        this.members = data.memberList?.map((mem) => new ClanMember_1.ClanMember(this.client, mem)) ?? []; // eslint-disable-line
    }
    /** Get {@link Player} info for every Player in the clan. */
    async fetchMembers(options) {
        return (await Promise.allSettled(this.members.map((m) => this.client.getPlayer(m.tag, { ...options, ignoreRateLimit: true }))))
            .filter((res) => res.status === 'fulfilled')
            .map((res) => res.value);
    }
    /** Get clan's formatted link to open clan in-game. */
    get shareLink() {
        return `https://link.clashofclans.com/en?action=OpenClanProfile&tag=${this.tag.replace(/#/g, '')}`;
    }
}
exports.Clan = Clan;
__decorate([
    (0, Decorators_1.Enumerable)(false)
], Clan.prototype, "client", void 0);
