// clashofclans.js/dist/struct/PlayerClan.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlayerClan = void 0;
const Decorators_1 = require("../util/Decorators");
const Badge_1 = require("./Badge");
/** Represents a Player's clan. */
class PlayerClan {
    constructor(client, data) {
        this.client = client;
        this.name = data.name;
        this.tag = data.tag;
        this.level = data.clanLevel ?? null; // eslint-disable-line
        this.badge = new Badge_1.Badge(data.badgeUrls);
    }
    /** Fetch detailed clan info for the player's clan. */
    fetch(options) {
        return this.client.getClan(this.tag, options);
    }
    /** Get clan's formatted link to open clan in-game. */
    get shareLink() {
        return `https://link.clashofclans.com/en?action=OpenClanProfile&tag=${this.tag.replace(/#/g, '')}`;
    }
}
exports.PlayerClan = PlayerClan;
__decorate([
    (0, Decorators_1.Enumerable)(false)
], PlayerClan.prototype, "client", void 0);
