// clashofclans.js/dist/struct/Ranking.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RankedClan = exports.RankedPlayer = exports.SeasonRankedPlayer = void 0;
const Constants_1 = require("../util/Constants");
const Badge_1 = require("./Badge");
const LeagueTier_1 = require("./LeagueTier");
const Location_1 = require("./Location");
const PlayerClan_1 = require("./PlayerClan");
/** Represents the Player of seasonal legend league leader-board ranking. */
class SeasonRankedPlayer {
    constructor(client, data) {
        this.name = data.name;
        this.tag = data.tag;
        this.rank = data.rank;
        this.expLevel = data.expLevel;
        this.trophies = data.trophies;
        this.attackWins = data.attackWins;
        this.defenseWins = data.defenseWins;
        // @ts-expect-error
        this.clan = data.clan ? new PlayerClan_1.PlayerClan(client, data.clan) : null;
    }
    /** Get player's formatted link to open player in-game. */
    get shareLink() {
        return `https://link.clashofclans.com/en?action=OpenPlayerProfile&tag=${this.tag.replace(/#/g, '')}`;
    }
}
exports.SeasonRankedPlayer = SeasonRankedPlayer;
/** Represents the Player of location based leader-board ranking. */
class RankedPlayer {
    constructor(client, data) {
        this.name = data.name;
        this.tag = data.tag;
        this.expLevel = data.expLevel;
        // @ts-expect-error
        this.trophies = data.trophies ?? null;
        // @ts-expect-error
        this.attackWins = data.attackWins ?? null;
        // @ts-expect-error
        this.defenseWins = data.defenseWins ?? null;
        // @ts-expect-error
        this.builderBaseTrophies = data.builderBaseTrophies ?? null;
        this.rank = data.rank;
        this.previousRank = data.previousRank ?? null; // eslint-disable-line
        // @ts-expect-error
        this.clan = data.clan ? new PlayerClan_1.PlayerClan(client, data.clan) : null;
        // @ts-expect-error
        this.leagueTier = data.trophies ? new LeagueTier_1.LeagueTier(data.leagueTier ?? Constants_1.UnrankedLeagueData) : null; // eslint-disable-line
    }
    /** Get player's formatted link to open player in-game. */
    get shareLink() {
        return `https://link.clashofclans.com/en?action=OpenPlayerProfile&tag=${this.tag.replace(/#/g, '')}`;
    }
}
exports.RankedPlayer = RankedPlayer;
/** Represents the Clan of location based leader-board ranking. */
class RankedClan {
    constructor(data) {
        this.name = data.name;
        this.tag = data.tag;
        this.level = data.clanLevel;
        // @ts-expect-error
        this.points = data.clanPoints ?? null;
        // @ts-expect-error
        this.builderBasePoints = data.clanBuilderBasePoints ?? null;
        this.location = new Location_1.Location(data.location);
        this.memberCount = data.members;
        this.rank = data.rank;
        this.previousRank = data.previousRank;
        this.badge = new Badge_1.Badge(data.badgeUrls);
    }
    /** Get clan's formatted link to open clan in-game. */
    get shareLink() {
        return `https://link.clashofclans.com/en?action=OpenClanProfile&tag=${this.tag.replace(/#/g, '')}`;
    }
}
exports.RankedClan = RankedClan;
