// clashofclans.js/dist/rest/HTTPError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrivateWarLogError = exports.HTTPError = void 0;
const messages = {
    500: 'Unknown error happened when handling the request.',
    504: 'The user aborted this request.',
    404: 'Requested resource was not found.',
    400: 'Client provided incorrect parameters for the request.',
    503: 'Service is temporarily unavailable because of maintenance.',
    429: 'Request was throttled, because amount of requests was above the threshold defined for the used API token.',
    403: 'Access denied, either because of missing/incorrect credentials or used API token does not grant access to the requested resource.'
};
const reasons = {
    503: 'inMaintenance',
    429: 'requestThrottled',
    400: 'badRequest',
    403: 'accessDenied',
    500: 'unknownException',
    404: 'notFound',
    504: 'requestAborted'
};
/** Represents an HTTP Error. */
class HTTPError extends Error {
    constructor(error, status, path, maxAge, method) {
        super();
        this.message = error?.message ?? messages[status];
        this.reason = error?.reason ?? reasons[status];
        this.path = path;
        this.method = method ?? 'GET';
        this.status = status;
        this.maxAge = maxAge;
    }
}
exports.HTTPError = HTTPError;
exports.PrivateWarLogError = {
    message: 'Access denied, clan war log is private.',
    reason: 'privateWarLog'
};
