// clashofclans.js/dist/rest/RequestHandler.js
"use strict";
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _RequestHandler_keyIndex;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequestHandler = void 0;
const node_events_1 = require("node:events");
const undici_1 = require("undici");
const Constants_1 = require("../util/Constants");
const Store_1 = require("../util/Store");
const Util_1 = require("../util/Util");
const HTTPError_1 = require("./HTTPError");
const IP_REGEX = /\d{1,3}.\d{1,3}.\d{1,3}.\d{1,3}/g;
/** Represents the class that manages handlers for endpoints. */
class RequestHandler extends node_events_1.EventEmitter {
    constructor(options) {
        super();
        _RequestHandler_keyIndex.set(this, 0); // eslint-disable-line
        this.onError = options?.onError;
        this.keys = options?.keys ?? [];
        this.retryLimit = options?.retryLimit ?? 0;
        this.throttler = options?.throttler ?? null;
        this.baseURL = options?.baseURL ?? Constants_1.APIBaseURL;
        this.restRequestTimeout = options?.restRequestTimeout ?? 0;
        this.rejectIfNotValid = options?.rejectIfNotValid ?? true;
        if (typeof options?.cache === 'object')
            this.cached = options.cache;
        else
            this.cached = options?.cache === true ? new Store_1.CacheStore() : null;
        this.dispatcher = new undici_1.Pool(new URL(this.baseURL).origin, {
            connections: options?.connections ?? null,
            pipelining: options?.pipelining ?? 1
        });
    }
    get _keys() {
        return Array.isArray(this.keys) ? this.keys : [this.keys];
    }
    get _key() {
        const key = this._keys[__classPrivateFieldGet(this, _RequestHandler_keyIndex, "f")];
        __classPrivateFieldSet(this, _RequestHandler_keyIndex, __classPrivateFieldGet(this, _RequestHandler_keyIndex, "f") + 1 >= this._keys.length ? 0 : __classPrivateFieldGet(this, _RequestHandler_keyIndex, "f") + 1, "f");
        return key;
    }
    setKeys(keys) {
        this.keys = keys;
        return this;
    }
    get credentials() {
        return Boolean(this.email && this.password);
    }
    async request(path, options = {}) {
        const cached = this.cached ? (await this.cached.get(path)) ?? null : null;
        if (cached && options.force !== true) {
            return {
                body: cached.body,
                res: { maxAge: cached.ttl - Date.now(), status: cached.status, path, ok: cached.status === 200 }
            };
        }
        if (!this.throttler || options.ignoreRateLimit)
            return this.exec(path, options);
        await this.throttler.wait();
        return this.exec(path, options);
    }
    async rawRequest(path, options = {}) {
        const cached = this.cached ? (await this.cached.get(path)) ?? null : null;
        if (cached && options.force !== true) {
            return cached.body;
        }
        if (!this.throttler || options.ignoreRateLimit)
            return this.exec(path, options).then((res) => res.body);
        await this.throttler.wait();
        const result = await this.exec(path, options);
        return result.body;
    }
    async exec(path, options = {}, retries = 0) {
        const result = await this.dispatch(path, options, retries);
        this.onError?.({ path, status: result.res.status, body: result.body });
        return result;
    }
    async dispatch(path, options = {}, retries = 0) {
        try {
            const res = await this.dispatcher.request({
                path: `/v1${path}`,
                body: options.body,
                method: options.method ?? 'GET',
                signal: (0, Util_1.timeoutSignal)(options.restRequestTimeout ?? this.restRequestTimeout),
                headers: { 'Authorization': `Bearer ${this._key}`, 'Content-Type': 'application/json' }
            });
            if (res.statusCode === 504 && retries < (options.retryLimit ?? this.retryLimit)) {
                return await this.exec(path, options, ++retries);
            }
            const body = (await res.body.json());
            if (this.credentials &&
                res.statusCode === 403 &&
                body?.reason === 'accessDenied.invalidIp' &&
                retries < (options.retryLimit ?? this.retryLimit)) {
                const keys = await this.reValidateKeys().then(() => this.login());
                if (keys.length)
                    return await this.exec(path, options, ++retries);
            }
            const maxAge = Number(res.headers['cache-control']?.split('=')?.[1] ?? 0) * 1000;
            if (res.statusCode === 403 && !body?.message && this.rejectIfNotValid) {
                throw new HTTPError_1.HTTPError(HTTPError_1.PrivateWarLogError, res.statusCode, path, maxAge);
            }
            if (res.statusCode !== 200 && this.rejectIfNotValid) {
                throw new HTTPError_1.HTTPError(body, res.statusCode, path, maxAge, options.method);
            }
            if (this.cached && maxAge > 0 && options.cache !== false && res.statusCode === 200) {
                await this.cached.set(path, { body, ttl: Date.now() + maxAge, status: res.statusCode }, maxAge);
            }
            return {
                body: body,
                res: { maxAge, status: res.statusCode, path, ok: res.statusCode === 200 }
            };
        }
        catch (error) {
            if (error.code === 'UND_ERR_ABORTED' && retries < (options.retryLimit ?? this.retryLimit)) {
                return this.exec(path, options, ++retries);
            }
            if (this.rejectIfNotValid)
                throw error;
            return {
                body: { message: error.message },
                res: { maxAge: 0, status: 500, path, ok: false }
            };
        }
    }
    async init(options) {
        if (!(options.email && options.password))
            throw ReferenceError('Missing email and password.');
        this.keyDescription = options.keyDescription;
        this.keyName = options.keyName ?? 'clashofclans.js.keys';
        this.keyCount = Math.min(options.keyCount ?? 1, 10);
        this.password = options.password;
        this.email = options.email;
        return this.reValidateKeys().then(() => this.login());
    }
    async reValidateKeys() {
        for (const key of this.keys) {
            const res = await (0, undici_1.fetch)(`${this.baseURL}/locations?limit=1`, {
                method: 'GET',
                signal: (0, Util_1.timeoutSignal)(this.restRequestTimeout),
                headers: { 'Authorization': `Bearer ${key}`, 'Content-Type': 'application/json' }
            }).catch(() => null);
            if (res?.status === 403) {
                const index = this.keys.indexOf(key);
                this.keys.splice(index, 1);
                process.emitWarning(`Key #${index + 1} is no longer valid. Removed from the key list.`);
            }
        }
    }
    async login() {
        const res = await (0, undici_1.fetch)(`${Constants_1.DevSiteAPIBaseURL}/login`, {
            method: 'POST',
            signal: (0, Util_1.timeoutSignal)(this.restRequestTimeout),
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ email: this.email, password: this.password })
        });
        const data = (await res.json());
        if (!res.ok)
            throw new Error(`Invalid email or password. ${JSON.stringify(data)}`);
        const ip = await this.getIp(data.temporaryAPIToken);
        if (!ip)
            throw new Error('Failed to get the IP address.');
        return this.getKeys(res.headers.get('set-cookie'), ip);
    }
    async getKeys(cookie, ip) {
        const res = await (0, undici_1.fetch)(`${Constants_1.DevSiteAPIBaseURL}/apikey/list`, {
            method: 'POST',
            signal: (0, Util_1.timeoutSignal)(this.restRequestTimeout),
            headers: { 'Content-Type': 'application/json', cookie }
        });
        const data = (await res.json());
        if (!res.ok)
            throw new Error(`Failed to retrieve the API Keys. ${JSON.stringify(data)}`);
        // Get all available keys from the developer site.
        const keys = (data.keys ?? []);
        // Revoke keys for specified key name but not matching current IP address.
        for (const key of keys.filter((key) => key.name === this.keyName && !key.cidrRanges?.includes(ip))) {
            if (!(await this.revokeKey(key.id, cookie)))
                continue;
            const index = keys.findIndex(({ id }) => id === key.id);
            keys.splice(index, 1);
        }
        // Filter keys for current IP address and specified key name.
        for (const key of keys.filter((key) => key.name === this.keyName && key.cidrRanges?.includes(ip))) {
            if (this.keys.length >= this.keyCount)
                break;
            if (!this.keys.includes(key.key))
                this.keys.push(key.key);
        }
        // Create keys within limits (maximum of 10 keys per account)
        while (this.keys.length < this.keyCount && keys.length < 10) {
            const key = await this.createKey(cookie, ip);
            this.keys.push(key.key);
            keys.push(key);
        }
        if (this.keys.length < this.keyCount && keys.length === 10) {
            process.emitWarning(`${this.keyCount} key(s) were requested but failed to create ${this.keyCount - this.keys.length} more key(s).`);
        }
        if (!this.keys.length) {
            throw new Error([
                `${keys.length} API keys were created but none match a key name of "${this.keyName}" and IP "${ip}".`,
                `Specify a key name or go to "https://developer.clashofclans.com" to delete unused keys.`
            ].join(' '));
        }
        return this.keys;
    }
    async revokeKey(keyId, cookie) {
        const res = await (0, undici_1.fetch)(`${Constants_1.DevSiteAPIBaseURL}/apikey/revoke`, {
            method: 'POST',
            signal: (0, Util_1.timeoutSignal)(this.restRequestTimeout),
            body: JSON.stringify({ id: keyId }),
            headers: { 'Content-Type': 'application/json', cookie }
        });
        return res.ok;
    }
    async createKey(cookie, ip) {
        const res = await (0, undici_1.fetch)(`${Constants_1.DevSiteAPIBaseURL}/apikey/create`, {
            method: 'POST',
            signal: (0, Util_1.timeoutSignal)(this.restRequestTimeout),
            headers: { 'Content-Type': 'application/json', cookie },
            body: JSON.stringify({
                cidrRanges: [ip],
                name: this.keyName,
                description: this.keyDescription ?? new Date().toUTCString()
            })
        });
        const data = (await res.json());
        if (!res.ok)
            throw new Error(`Failed to create API Key. ${JSON.stringify(data)}`);
        return data.key;
    }
    async getIp(token) {
        try {
            const decoded = JSON.parse(Buffer.from(token.split('.')[1], 'base64').toString());
            const props = decoded.limits.find((limit) => limit.hasOwnProperty('cidrs'));
            return props.cidrs[0].match(IP_REGEX)[0];
        }
        catch {
            const body = await (0, undici_1.fetch)('https://api.ipify.org', {
                signal: (0, Util_1.timeoutSignal)(this.restRequestTimeout)
            }).then((res) => res.text());
            return body.match(IP_REGEX)?.[0] ?? null;
        }
    }
}
exports.RequestHandler = RequestHandler;
_RequestHandler_keyIndex = new WeakMap();
