// clashofclans.js/dist/util/Constants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RawData = exports.CWLRounds = exports.RestEvents = exports.ClientEvents = exports.PollingEvents = exports.FriendlyWarPreparationTimes = exports.WarLeagues = exports.Leagues = exports.UnrankedWarLeagueId = exports.LegendLeagueId = exports.UnrankedLeagueId = exports.UnrankedLeagueData = exports.HeroPets = exports.Heroes = exports.BuilderTroops = exports.Spells = exports.DarkElixirSpells = exports.ElixirSpells = exports.HeroEquipment = exports.SiegeMachines = exports.HomeTroops = exports.DarkElixirTroops = exports.ElixirTroops = exports.SuperTroops = exports.DevSiteAPIBaseURL = exports.APIBaseURL = void 0;
const raw_json_1 = __importDefault(require("../util/raw.json"));
exports.APIBaseURL = 'https://api.clashofclans.com/v1';
exports.DevSiteAPIBaseURL = 'https://developer.clashofclans.com/api';
exports.SuperTroops = raw_json_1.default.RAW_SUPER_UNITS.map((unit) => unit.name);
exports.ElixirTroops = raw_json_1.default.RAW_UNITS.filter((unit) => !exports.SuperTroops.includes(unit.name) && unit.subCategory === 'troop' && unit.upgrade.resource === 'Elixir').map((unit) => unit.name);
exports.DarkElixirTroops = raw_json_1.default.RAW_UNITS.filter((unit) => !exports.SuperTroops.includes(unit.name) && unit.subCategory === 'troop' && unit.upgrade.resource === 'Dark Elixir').map((unit) => unit.name);
exports.HomeTroops = [...exports.ElixirTroops, ...exports.DarkElixirTroops];
exports.SiegeMachines = raw_json_1.default.RAW_UNITS.filter((unit) => unit.subCategory === 'siege').map((unit) => unit.name);
exports.HeroEquipment = raw_json_1.default.RAW_UNITS.filter((unit) => unit.subCategory === 'equipment').map((unit) => unit.name);
exports.ElixirSpells = raw_json_1.default.RAW_UNITS.filter((unit) => unit.subCategory === 'spell' && unit.upgrade.resource === 'Elixir').map((unit) => unit.name);
exports.DarkElixirSpells = raw_json_1.default.RAW_UNITS.filter((unit) => unit.subCategory === 'spell' && unit.upgrade.resource === 'Dark Elixir').map((unit) => unit.name);
exports.Spells = [...exports.ElixirSpells, ...exports.DarkElixirSpells];
exports.BuilderTroops = raw_json_1.default.RAW_UNITS.filter((unit) => unit.subCategory === 'troop' && unit.village === 'builderBase').map((unit) => unit.name);
exports.Heroes = raw_json_1.default.RAW_UNITS.filter((unit) => unit.subCategory === 'hero').map((unit) => unit.name);
exports.HeroPets = raw_json_1.default.RAW_UNITS.filter((unit) => unit.subCategory === 'pet').map((unit) => unit.name);
exports.UnrankedLeagueData = {
    id: 105000000,
    name: 'Unranked',
    iconUrls: {
        small: 'https://api-assets.clashofclans.com/leaguetiers/125/yyYo5DUFeFBZvmMEQh0ZxvG-1sUOZ_S3kDMB7RllXX0.png',
        large: 'https://api-assets.clashofclans.com/leaguetiers/326/yyYo5DUFeFBZvmMEQh0ZxvG-1sUOZ_S3kDMB7RllXX0.png'
    }
};
exports.UnrankedLeagueId = 29000000;
exports.LegendLeagueId = 29000022;
exports.UnrankedWarLeagueId = 48000000;
exports.Leagues = [
    29000000,
    29000001,
    29000002,
    29000003,
    29000004,
    29000005,
    29000006,
    29000007,
    29000008,
    29000009,
    29000010,
    29000011,
    29000012,
    29000013,
    29000014,
    29000015,
    29000016,
    29000017,
    29000018,
    29000019,
    29000020,
    29000021,
    exports.LegendLeagueId
];
exports.WarLeagues = [
    48000000, 48000001, 48000002, 48000003, 48000004, 48000005, 48000006, 48000007, 48000008, 48000009, 48000010, 48000011, 48000012,
    48000013, 48000014, 48000015, 48000016, 48000017, 48000018
];
exports.FriendlyWarPreparationTimes = [
    1000 * 60 * 60 * 24,
    1000 * 60 * 60 * 20,
    1000 * 60 * 60 * 16,
    1000 * 60 * 60 * 12,
    1000 * 60 * 60 * 8,
    1000 * 60 * 60 * 6,
    1000 * 60 * 60 * 4,
    1000 * 60 * 60 * 2,
    1000 * 60 * 60,
    1000 * 60 * 30,
    1000 * 60 * 15,
    1000 * 60 * 5
];
exports.PollingEvents = {
    ClanLoopStart: 'clanLoopStart',
    ClanLoopEnd: 'clanLoopEnd',
    PlayerLoopStart: 'playerLoopStart',
    PlayerLoopEnd: 'playerLoopEnd',
    WarLoopStart: 'warLoopStart',
    WarLoopEnd: 'warLoopEnd',
    NewSeasonStart: 'newSeasonStart',
    MaintenanceStart: 'maintenanceStart',
    MaintenanceEnd: 'maintenanceEnd',
    Error: 'error',
    Debug: 'debug'
};
exports.ClientEvents = {
    Error: 'error',
    Debug: 'debug'
};
exports.RestEvents = {
    Error: 'error',
    Debug: 'debug',
    RateLimited: 'rateLimited'
};
exports.CWLRounds = {
    PreviousRound: 'warEnded',
    CurrentRound: 'inWar',
    NextRound: 'preparation'
};
exports.RawData = {
    RawUnits: raw_json_1.default.RAW_UNITS,
    RawSuperUnits: raw_json_1.default.RAW_SUPER_UNITS
};
